File: Practica_1/src/gestorAplicacion/gestion/Pedido.java

package gestorAplicacion.gestion;

import gestorAplicacion.usuarios.Cliente;
import gestorAplicacion.usuarios.Gerente;
import gestorAplicacion.usuarios.Trabajador;

import java.io.Serializable;
import java.time.LocalDate;
import java.util.*;
/*
* Clase creada para almacenar los prodcutos deseados por el cliente
* Tambien puede contener servicios
* El pedido esta asociado tanto a trabajador como al cliente
* */


public class Pedido implements Serializable {
    private static final long serialVersionUID = 1L;

    //ATRIBUTOS DE INSTANCIA

    private Trabajador trabajador;
    private Cliente cliente;
    private String estadoPedido; // "Pagado", "No pagado"
    private ArrayList<Producto> productos = new ArrayList<>();//arraylist de productos pedidos por el cliente

    private ArrayList<Servicio> servicios =new ArrayList<>();//arraylist de servicios pedidos por el cliente
    private LocalDate fechaPedido;
    private long codigo; //cuando se cree, se cree con el mismo ID del cliente

    private static int totalPedidos = 0;

    //ATRIBUTOS DE CLASE

    private static ArrayList<Pedido> pedidos= new ArrayList<>();
    private static long numeroPedido=0;

    //CONSTRUCTOR


    public Pedido() {
    }

    public Pedido(Cliente cliente, String estadoPedido, ArrayList<Producto> productos, ArrayList<Servicio> servicios, LocalDate fechaPedido, long codigo) {
        this.cliente = cliente;
        this.estadoPedido = estadoPedido; //CONSTRUCTOR PARA LA FUNCIONALIDAD realizarReserva
        this.productos = productos;
        this.servicios = servicios;
        this.fechaPedido = fechaPedido;
        this.codigo = codigo;
        pedidos.add(this);
    }

    public Pedido(Trabajador trabajador, Cliente cliente, String estadoPedido, ArrayList<Producto> productos, ArrayList<Servicio> servicios, LocalDate fechaPedido) {
        this.trabajador = trabajador;
        this.cliente = cliente;
        this.estadoPedido = estadoPedido;
        this.productos = productos;
        this.servicios = servicios;
        this.fechaPedido = fechaPedido;
        this.codigo = generarCodigo();
        pedidos.add(this);
        Pedido.totalPedidos += 1;
    }

    //GETTERS Y SETTERS


    public Trabajador getTrabajador() {
        return trabajador;
    }

    public void setTrabajador(Trabajador trabajador) {
        this.trabajador = trabajador;
    }

    public Cliente getCliente() {
        return cliente;
    }

    public void setCliente(Cliente cliente) {
        this.cliente = cliente;
    }

    public static ArrayList<Pedido> getPedidos() {
        return pedidos;
    }

    public static void setPedidos(ArrayList<Pedido> pedidos) {
        Pedido.pedidos = pedidos;
    }

    public String getEstadoPedido() {
        return estadoPedido;
    }

    public static int getTotalPedidos() {
        return pedidos.size();
    }

    public static void setTotalPedidos(int totalPedidos) {
        Pedido.totalPedidos = totalPedidos;
    }


    public void setEstadoPedido(String estadoPedido) {
        this.estadoPedido = estadoPedido;
    }

    public ArrayList<Producto> getProductos() {
        return productos;
    }

    public void setProductos(ArrayList<Producto> productos) {
        this.productos = productos;
    }

    public LocalDate getFechaPedido() {
        return fechaPedido;
    }

    public void setFechaPedido(LocalDate fechaPedido) {
        this.fechaPedido = fechaPedido;
    }

    public long getCodigo() {
        return codigo;
    }

    public void setCodigo(int id) {
        this.codigo = codigo;
    }

    public static long getNumeroPedido() {
        return numeroPedido;
    }

    public static void setNumeroPedido(long numeroPedido) {
        Pedido.numeroPedido = numeroPedido;
    }

    public ArrayList<Servicio> getServicios() {
        return servicios;
    }

    public void setServicios(ArrayList<Servicio> servicios) {
        this.servicios = servicios;
    }

    public void setCodigo(long codigo) {
        this.codigo = codigo;
    }

    //OTROS METODOS
    public long generarCodigo(){
        return numeroPedido++;
    }



    public String generarFactura(){ //genera la factura del pedido realizado
        String productos ="Productos: " +
                "\nNombre........Cantidad.......Precio";
        HashMap<String, Double> nombreYprecio = new HashMap<>();
        double totalProductos = 0;
        ArrayList<String> nombre = new ArrayList<>();
        for (Producto producto : this.getProductos()) {
            if (nombreYprecio.containsKey(producto.getNombre()) == false) {
                nombreYprecio.put(producto.getNombre(), producto.getPrecioVenta());
            }
            nombre.add(producto.getNombre());
            totalProductos += producto.getPrecioVenta();
        }
        long i = 1;
        for (String clave : nombreYprecio.keySet()) {
            productos += "\n" + i + ". Nombre: " + clave + ", Cantidad pedida: " + Collections.frequency(nombre, clave) + "......" + "Precio: "+nombreYprecio.get(clave) * Collections.frequency(nombre, clave);
            i++;
        }

        productos += "\nTotal de producto:" + totalProductos;

        String servicios="Servicios:" +
                "\nNombre...........Cantidad.........Precio";
        int b=1;
        double totalServicios=0;
        SortedSet<Servicio> servicioNoRepetido=new TreeSet<>();
        for(Servicio servicio: this.getServicios()){
            if(servicioNoRepetido.contains(servicio)==false){
                servicios+="\n"+b+". Servicio: "+servicio+", Cantidad: "+ Collections.frequency(this.getServicios(),servicio)+"....."+"Precio: "+Collections.frequency(this.getServicios(),servicio)* servicio.getPrecio();
                b++;
            }
            totalServicios+=servicio.getPrecio();
            servicioNoRepetido.add(servicio);

        }
        servicios += "\nTotal de servicios:" + totalServicios;


        String s="";
        s+= "\n=============FACTURA DEL PEDIDO============="+
                "\nFactura # "+this.getCodigo()+
                "\nFecha: "+this.getFechaPedido()+
                "\nNombre cliente: "+this.getCliente().getNombre()+
                "\nVendido por "+this.getTrabajador().getNombre()+" con codigo "+this.getTrabajador().getCodigo()+
                "\n"+productos+
                "\n"+servicios+
                "\nTotal:"+ (totalProductos+totalServicios)+
                "\n===========================================";
        return s;
    }

    public double sumaProductos(){ // calcula el precio del total de los productos pedidos
        double sumaTotal=0;
        for (Producto producto: this.getProductos()){
            sumaTotal+=producto.getPrecioVenta();
        }
        return sumaTotal;
    }
    public double sumaServicios(){ // calcula el precio del total de los servicios pedidos
        double sumaTotal=0;
        for (Servicio servicio: this.getServicios()){
            sumaTotal+=servicio.getPrecio();
        }
        return sumaTotal;
    }
    public static String mostraPedidos(){ // Muestra los pedidos no pagados
        String s="";
        long i=1;
        for(Pedido pedido: Pedido.getPedidos()){
            if(pedido.getEstadoPedido().equals("No pagado")){
                double productos=pedido.sumaProductos();
                double servicios=pedido.sumaServicios();
                s+="\n"+"Codigo de pedido: "+pedido.getCodigo()+", Cliente: "+ pedido.getCliente().getNombre()+", Valor total: "+(productos+servicios);
                i++;
            }
        }
        return s;
    }
    public static Pedido seleccionarPedido(long codigo){ // ingresa el codigo del pedido y retorna el objeto pedido correspondiente
        Pedido pedido1=null;
        for(Pedido pedido: Pedido.getPedidos()){
            if(pedido.getCodigo()==codigo){
                pedido1=pedido;
                break;
            }
        }
        return pedido1;
    }

}

File: Python/gestorAplicacion_py/gestion/Pedido.py


#
#* Clase creada para almacenar los prodcutos deseados por el cliente
#* Tambien puede contener servicios
#* El pedido esta asociado tanto a trabajador como al cliente
#* 


class Pedido:




    #ATRIBUTOS DE INSTANCIA



    _totalPedidos = 0

    #ATRIBUTOS DE CLASE

    _pedidos = []
    _numeroPedido =0

    #CONSTRUCTOR

    def __init__(self, trabajador, cliente, estadoPedido="", productos=0, servicios=[], fechaPedido="", codigo=0):
        self._cliente = cliente
        self._trabajador = trabajador
        self._estadoPedido = estadoPedido #CONSTRUCTOR PARA LA FUNCIONALIDAD realizarReserva
        self._producto= productos
        self._productos = [] #array con productos
        self._servicios = servicios  #array con servicios
        self._fechaPedido = fechaPedido #fecha del pedido
        self._codigo = Pedido._totalPedidos        #codigo del pedido
        Pedido._pedidos.append(self)
        Pedido._totalPedidos+=1




    #GETTERS Y SETTERS


    def getTrabajador(self):
        return self._trabajador

    def setTrabajador(self, trabajador):
        self._trabajador = trabajador

    def getCliente(self):
        return self._cliente

    def setCliente(self, cliente):
        self._cliente = cliente

    
    @classmethod
    def getPedidos(cls):
        return cls._pedidos

    @classmethod
    def setPedidos(cls, pedidos):
        cls._pedidos = pedidos


    @classmethod
    def getTotalPedidos(cls):
        return cls._totalPedidos

    @classmethod
    def setTotalPedidos(cls, totalPedidos):
        cls._totalPedidos = totalPedidos

    def setEstadoPedido(self, estadoPedido):
        self._estadoPedido = estadoPedido

    def getProductos(self):
        return self._productos

    def setProductos(self, productos):
        self._productos = productos

    def getFechaPedido(self):
        return self._fechaPedido

    def setFechaPedido(self, fechaPedido):
        self._fechaPedido = fechaPedido

    def getCodigo(self):
        return self._codigo

    def setCodigo(self, codigo):
        self._codigo = codigo


    def getServicios(self):
        return self._servicios

    def setServicios(self, servicios):
        self._servicios = servicios

    def setCodigo(self, codigo):
        self.codigo = codigo

    @classmethod
    def getNumeroPedido(cls):
        return cls._numeroPedido

    @classmethod
    def setNumeroPedido(cls, numeroPedido):
        cls._numeroPedido = numeroPedido



#OTROS METODOS






    def sumaProductos(self):
        sumaTotal =0
        for producto in self.getProductos():
            sumaTotal+=producto.getPrecioVenta()
        return sumaTotal

    def sumaServicios(self):
        sumaTotal =0
        for servicio in self.getServicios():
            sumaTotal+=servicio.getPrecio()
        return sumaTotal
    @staticmethod
    def mostraPedidos():
        s =""
        i =1
        for pedido in Pedido.getPedidos():
            if pedido.getEstadoPedido() is "No pagado":
                productos =pedido.sumaProductos()
                servicios =pedido.sumaServicios()
                s+="\n"+"Codigo de pedido: "+pedido.getCodigo()+", Cliente: "+ pedido.getCliente().getNombre()+", Valor total: "+str(productos+servicios)
                i += 1
        return s

    @staticmethod
    def seleccionarPedido(codigo):
        pedido1 =None
        for pedido in Pedido.getPedidos():
            if pedido.getCodigo()==codigo:
                pedido1=pedido
                break
        return pedido1



    


    